### PR TITLE
bloat: init at unstable-2022-03-31

### PR DESCRIPTION
**Pull request by fgaz**
###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [x] (Package updates) Added a release notes entry if the change is major or breaking
  - [x] (Module updates) Added a release notes entry if the change is significant
  - [x] (Module addition) Added a release notes entry if adding a new NixOS module
  - [x] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/servers/bloat/default.nix [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+, fetchgit
+, unstableGitUpdater
+}:
+
+buildGoModule {
+  pname = "bloat";
+  version = "unstable-2022-02-11";
+
+  src = fetchgit {
+    url = "git://git.freesoftwareextremist.com/bloat";
```
Comment (papojari) [inline pkgs/servers/bloat/default.nix:12]:
```suggestion
    url = "https://git.freesoftwareextremist.com/bloat";
```
`git://` is considered unsafe.

File: pkgs/top-level/all-packages.nix [hunk 2]
```diff
@@ -21027,6 +21027,8 @@ with pkgs;
 
   bird = callPackage ../servers/bird { };
 
+  bloat = callPackage ../servers/bloat { };
```
Comment (papojari) [inline pkgs/top-level/all-packages.nix:21030]: The description says that this is a client. So why `servers`?

```suggestion
  bloat = callPackage ../applications/networking/instant-messengers/bloat { };
```
would be better. Other Mastodon clients are also in that catergory.

You'd still have to move the `default.nix`.

**Review by auroraanna — CHANGES_REQUESTED** [review]

**Comment by papojari** [comment]
Also:
```
[nix-shell:~/.cache/nixpkgs-review/pr-162082]$ ./results/bloat/bin/bloat 
open /etc/bloat.conf: no such file or directory
```

File: pkgs/servers/bloat/default.nix [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+    rev = "c390a0c32720f6afe852bc7a3a3f64c3afe9e401";
+    sha256 = "sha256-SeIpm2zDnjllnsr+It5rIem+WNKRoJH1WqJuEL2hkRo=";
+  };
+
+  vendorSha256 = null;
```
Comment (papojari) [inline pkgs/servers/bloat/default.nix:17]: Why no `vendorSha256`?

**Review by auroraanna — CHANGES_REQUESTED** [review]

**Comment by papojari** [comment]
##### Reviewed points

- [ ] package path fits guidelines
- [X] package name fits guidelines
- [X] package version fits guidelines
- [X] package build on `x86_64-linux`
- [X] executables tested on `x86_64-linux`
- [X] `meta.description` is set and fits guidelines
- [X] `meta.license` fits upstream license
- [X] `meta.platforms` is set
- [X] `meta.maintainers` is set
- [X] build time only dependencies are declared in `nativeBuildInputs`
- [ ] source is fetched using the appropriate function
- [X] phases are respected
- [X] patches that are remotely available are fetched with `fetchpatch`

##### Possible improvements

All the previous comments.

##### Comments

File: pkgs/top-level/all-packages.nix (hunk 2 above)
Comment (fgaz) [reply pkgs/top-level/all-packages.nix:21030]: Well it's kind of both! It is a pleroma/mastodon client, but it runs a web server that serves the frontend, and it's mostly set up by server operators rather than by users themselves

File: pkgs/servers/bloat/default.nix (hunk 3 above)
Comment (fgaz) [reply pkgs/servers/bloat/default.nix:17]: https://nixos.org/manual/nixpkgs/stable/index.html#ssec-language-go

> vendorSha256 can also take null as an input. When null is used as a value, rather than fetching the dependencies and vendoring them, we use the vendoring included within the source repo

File: pkgs/servers/bloat/default.nix (hunk 1 above)
Comment (fgaz) [reply pkgs/servers/bloat/default.nix:12]: Yeah but we have the hash, and that is dumb http, so eg. --depth does not work and the whole repo has to be fetched. I guess I could also fetch a tarball, but that'd break unstableGitUpdater :-/

I'm not sure what's the best option here

File: pkgs/top-level/all-packages.nix (hunk 2 above)
Comment (papojari) [reply pkgs/top-level/all-packages.nix:21030]: The client may use a web server but it's still a client and not serving other users, right? So it shouldn't be in the server category.

File: pkgs/servers/bloat/default.nix (hunk 1 above)
Comment (papojari) [reply pkgs/servers/bloat/default.nix:12]: Why would the whole repo not have to be fetched? What is your use of `--depth` here?

Comment (fgaz) [reply pkgs/servers/bloat/default.nix:12]: > Why would the whole repo not have to be fetched?

I meant the whole history

> What is your use of --depth here?

unstableGitUpdater only needs HEAD, so it uses `--depth 1`, and I assume that fetchgit does the same

File: pkgs/top-level/all-packages.nix (hunk 2 above)
Comment (fgaz) [reply pkgs/top-level/all-packages.nix:21030]: > The client may use a web server but it's still a client and not serving other users, right?

It's made to serve multiple users. eg, https://bloat.freesoftwareextremist.com/ is meant to be used by any user of freesoftwareextremist.com (or any other mastodon / pleroma instance for that matter)

File: pkgs/servers/bloat/default.nix (hunk 1 above)
Comment (papojari) [reply pkgs/servers/bloat/default.nix:12]: is writing at least disabled over `git://`. If so I'd go with it.

File: pkgs/top-level/all-packages.nix (hunk 2 above)
Comment (papojari) [reply pkgs/top-level/all-packages.nix:21030]: So It'd be nice to have a module to enable the server, right?

Comment (fgaz) [reply pkgs/top-level/all-packages.nix:21030]: Yes I do plan to write one, just not in this pr

File: pkgs/servers/bloat/default.nix [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+  meta = with lib; {
+    description = "A web client for Pleroma and Mastodon";
+    longDescription = ''
+      A lightweight web client for Pleroma and Mastodon.
+      Does not require JavaScript to display text, images, audio and videos.
+    '';
+    homepage = "https://bloat.freesoftwareextremist.com";
+    downloadPage = "https://git.freesoftwareextremist.com/bloat/";
+    license = licenses.cc0;
+    platforms = platforms.all;
```
Comment (SuperSandro2000) [inline pkgs/servers/bloat/default.nix:40]:
```suggestion
```
set by buildGoModule to darwin+linux.

Comment (fgaz) [reply pkgs/servers/bloat/default.nix]: Done

File: pkgs/servers/bloat/default.nix (hunk 1 above)
Comment (papojari) [reply pkgs/servers/bloat/default.nix:12]: I mean pushing to it

Comment (fgaz) [reply pkgs/servers/bloat/default.nix:12]: I have no idea, but only the owners are going to use it for writing, not nixpkgs users

File: pkgs/servers/bloat/default.nix [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+
+  postInstall = ''
+    mkdir -p $out/share/bloat
+    cp -r templates $out/share/bloat/templates
+    cp -r static $out/share/bloat/static
+    sed \
+      -e "s%=templates%=$out/share/bloat/templates%g" \
+      -e "s%=static%=$out/share/bloat/static%g"       \
+      < bloat.conf > $out/share/bloat/bloat.conf.example
+  '';
```
Comment (papojari) [inline pkgs/servers/bloat/default.nix:27]: The config files are being copied to the output directory but how are you supposed to access them without knowing the directory name of the derivation?

Comment (fgaz) [reply pkgs/servers/bloat/default.nix:27]: There are multiple ways to know the path of a derivation. Though I don't expect people to actually use that file. It's there mostly because the makefile generates it too but with fixed paths, and to make it easier to run the server until I write a proper module

Comment (papojari) [reply pkgs/servers/bloat/default.nix:27]: What that config is the default config? You should be able to use the package without the module.

Comment (fgaz) [reply pkgs/servers/bloat/default.nix:27]: > What that config is the default config?

I don't understand your question, could you rephrase?

> You should be able to use the package without the module.

It is possible. Just [create a configuration](https://git.freesoftwareextremist.com/bloat/tree/README#n18) (whether you pick the template from upstream or from the store path is irrelevant) and run the program

Comment (papojari) [reply pkgs/servers/bloat/default.nix:27]: > I don't understand your question, could you rephrase?

I think I meant something along the lines of why don't you make the nix store config the default config. If you want to change the config you can then make you own config. I mean you already put the config into the nix store so why not make it a bit easier to use and make it the default config?

Comment (fgaz) [reply pkgs/servers/bloat/default.nix:27]: I _could_ do that, and it would be as simple as running `wrapProgram $out/bin/bloat --add-flags "-f $out/share/bloat/bloat.conf.example"` in `postInstall`, _and_ the user could even easily override it since only the last `-f` is used, but I think that's going a bit too far. This also prevents running the program without arguments to see the usage text. @SuperSandro2000 what do you think?

Comment (papojari) [reply pkgs/servers/bloat/default.nix:27]: If you can't pull up the help info without starting the it that's a problem. Just leave it like this then.

Comment (SuperSandro2000) [reply pkgs/servers/bloat/default.nix:27]: Aren't those files linked `/var/share/bloat`? Should be good enough.

**Comment by fgaz** [comment]
Is there anything else I need to do?
